File: scripts/day10-part1-simple.py
import time
import re

def parse_machine(line):
    match = re.match(r'\[(.*?)\](.*)\{(.*)\}', line)
    lights = match.group(1)
    buttons_str = match.group(2).strip()
    joltages = list(map(int, match.group(3).split(',')))
    
    buttons = []
    for button_match in re.finditer(r'\(([0-9,]+)\)', buttons_str):
        buttons.append(list(map(int, button_match.group(1).split(','))))
    
    return {'lights': lights, 'buttons': buttons, 'joltages': joltages}

def solve_part1(machines):
    total = 0
    
    for machine in machines:
        target = [1 if c == '#' else 0 for c in machine['lights']]
        n = len(target)
        m = len(machine['buttons'])
        
        # Try all combinations
        min_presses = float('inf')
        for mask in range(1 << m):
            state = [0] * n
            presses = 0
            
            for i in range(m):
                if mask & (1 << i):
                    presses += 1
                    for light in machine['buttons'][i]:
                        state[light] ^= 1
            
            if state == target:
                min_presses = min(min_presses, presses)
        
        total += min_presses
    
    return total

# Read input
with open("../inputs/day-10-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

machines = [parse_machine(line) for line in input_data]

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part1(machines)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 10 Part 1 - Simple)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")

File: scripts/day10-part1-simple.js
const fs = require("fs");

function parseMachine(line) {
  const match = line.match(/\[(.*?)\](.*)\{(.*)\}/);
  const lights = match[1];
  const buttonsStr = match[2].trim();
  const joltages = match[3].split(",").map(Number);

  const buttons = [];
  const buttonMatches = buttonsStr.matchAll(/\(([0-9,]+)\)/g);
  for (const m of buttonMatches) {
    buttons.push(m[1].split(",").map(Number));
  }

  return { lights, buttons, joltages };
}

function solvePart1(machines) {
  let total = 0;

  for (const machine of machines) {
    const target = machine.lights.split("").map((c) => (c === "#" ? 1 : 0));
    const n = target.length;
    const m = machine.buttons.length;

    // Try all combinations (brute force for simple version)
    let minPresses = Infinity;
    for (let mask = 0; mask < 1 << m; mask++) {
      const state = new Array(n).fill(0);
      let presses = 0;

      for (let i = 0; i < m; i++) {
        if (mask & (1 << i)) {
          presses++;
          for (const light of machine.buttons[i]) {
            state[light] ^= 1;
          }
        }
      }

      if (state.every((s, i) => s === target[i])) {
        minPresses = Math.min(minPresses, presses);
      }
    }

    total += minPresses;
  }

  return total;
}

const input = fs
  .readFileSync("../inputs/day-10-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const machines = input.map(parseMachine);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart1(machines);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 10 Part 1 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
